engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java

```diff
@@ -0,0 +1,189 @@
+     *
+     * @return this node's children.
+     */
+    List<IR> children();
+
+    /**
+     * Lists all the nodes in the preorder walk of the tree of this node.
+     *
+     * @return all the descendants of this node.
+     */
+    default List<IR> preorder() {
+        List<IR> ordered = children().flatMap(c -> c.preorder());
+        IR element = this;
+        return $colon$colon$.MODULE$.apply(element, ordered); // ordered.prepended(element) is reporeted as ambiguous
+    }
+
+    /**
+     * Pretty prints the IR.
+     *
+     * @return a pretty-printed representation of the IR
+     */
+    default String pretty() {
+        return Debug.pretty(this.toString());
+    }
+
+    /**
+     * Gets the node's identifier.
+     *
+     * @return the node's identifier
+     */
+    default IR.Identifier getId() {
+        return id();
+    }
+
+    /**
+     * A unique identifier for a piece of IR.
+     */
+    IR.Identifier id();
+
+    /**
+     * Storage for compiler diagnostics related to the IR node.
+     */
+    DiagnosticStorage diagnostics();
+
+    /**
+     * Creates a deep structural copy of `this`, representing the same structure.
+     * <p>
+     * You can choose to keep the location, metadata and diagnostic information
+     * in the duplicated copy, as well as whether you want to generate new
+     * node identifiers or not.
+     *
+     * @param keepLocations   whether locations should be kept in the duplicated IR
+     * @param keepMetadata    whether the pass metadata should be kept in the duplicated IR
+     * @param keepDiagnostics whether the diagnostics should be kept in the duplicated IR
+     * @param keepIdentifiers whether the identifiers should be regenerated in the duplicated IR
+     * @return a deep structural copy of `this`
+     */
+    IR duplicate(
+            boolean keepLocations,
+            boolean keepMetadata,
+            boolean keepDiagnostics,
+            boolean keepIdentifiers
+    );
+
+    /**
+     * Shows the IR as code.
+     *
+     * @param indent the current indentation level
+     * @return a string representation of `this`
+     */
+    String showCode(int indent);
+
+    default String showCode() {
+        return showCode(0);
+    }
+
+
+    /** Creates a random identifier.
+     *
+     * @return a random identifier
+     */
+    static IR.Identifier randomId() {
+        return new Identifier(UUID.randomUUID());
+    }
+
+    /** The type of identifiers for IR nodes. */
+    record Identifier(UUID id) implements Serializable {}
+
+
+    /** The type of external identifiers */
+    record ExternalId(UUID id) implements Serializable {}
+
+    /** Generates an indent of `n` spaces.
+     *
+     * @param n the number of spaces
+     * @return a string representing an `n`-space indent
+     */
+    static String mkIndent(int n) {
+        return " ".repeat(n);
+    }
+
+    /** The size of a single indentation level. */
+    int indentLevel = 4;
```

Comment:
I'd prefixed the field with `public static final` - but maybe that's the default.

Reply:
That's the default, I just followed IDE suggestion to remove it.